src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
+                      // If it is a nameof, skip the 'if' and parse as an expression. 
+                      (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken))
                 {
-                    TypeSyntax type = this.ParseType(ParseTypeMode.AfterCase);
-                    if (!type.IsMissing)
+                    type = this.ParseType(ParseTypeMode.DefinitePattern);
+                    if (type.IsMissing)
                     {
-                        // X.Y.Z id
-                        if (this.IsTrueIdentifier() && this.CurrentToken.ContextualKind != SyntaxKind.WhenKeyword)
-                        {
-                            var designation = ParseSimpleDesignation();
-                            node = _syntaxFactory.DeclarationPattern(type, designation);
-                        }
+                        // either it is not shaped like a type, or it is a constant expression.
+                        this.Reset(ref resetPoint);
+                        type = null;
+                    }
+                }
+
+                PropertySubpatternSyntax propertySubpattern = null;
+                bool parsePropertySubpattern()
+                {
+                    if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken)
+                    {
+                        propertySubpattern = ParsePropertySubpattern();
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                VariableDesignationSyntax designation = null;
+                bool parseDesignation()
+                {
+                    if (this.IsTrueIdentifier() && (!forCase || this.CurrentToken.ContextualKind != SyntaxKind.WhenKeyword))
+                    {
+                        designation = ParseSimpleDesignation();
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                bool looksLikeCast()
+                {
+                    var resetPoint2 = this.GetResetPoint();
+                    try
+                    {
+                        return this.ScanCast(forPattern: true);
                     }
-                    if (node == null)
+                    finally
                     {
-                        // it is an expression for typical switch case. 
+                        this.Reset(ref resetPoint2);
+                        this.Release(ref resetPoint2);
+                    }
+                }
+
+                if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken && (type != null || !looksLikeCast()))
+                {
+                    // It is possible this is a parenthesized (constant) expression.
+                    // We normalize later.
+                    ParseSubpatternList(out var openParenToken, out var subPatterns, out var closeParenToken, SyntaxKind.OpenParenToken, SyntaxKind.CloseParenToken);
```

Comment:
>ParseSubpatternList(out var openParenToken, out var subPatterns, out var closeParenToken, SyntaxKind.OpenParenToken, SyntaxKind.CloseParenToken); [](start = 20, length = 145)

Please spell out the types.

Reply:
Recorded in [https://github.com/dotnet/roslyn/issues/23222](https://github.com/dotnet/roslyn/issues/23222).

---
In reply to: [151812397](https://github.com/dotnet/roslyn/pull/23208#discussion_r151812397) [](ancestors = 151812397)